### ISSUE TITLE
Mix Tab function between win and unix 

### ISSUE THREAD

**Issue by mrx23dot** [chrisant996/clink#166]
With `clink set exec.enable on`
only executables are listed on Tab press, this prevents listing "new text doc.txt", but useful for completing like pyt to python.

with `clink set exec.enable off`
Folder content is listed, but there is no executable completion.

Could we have it both?

**Comment by chrisant996** [comment]
It already has that.

As mentioned in the Cmder issue, please refer to the Clink docs, regarding the multiple different settings whose names begin with `exec.`.

Chose the combination of them that you prefer, regarding which kinds of things to expand for the first word on a command line.

Since you're using Cmder, the docs are in `vendor\clink\clink.html` under the Cmder directory.  Look for the section on Configuring Clink, and Clink Settings.

Or online docs, e.g. https://chrisant996.github.io/clink/clink.html#clink-settings

**Comment by mrx23dot** [comment]
I tried setting both at the same time, but excutables always seem to suppress normal files with this combo:
clink set exec.enable on
clink set exec.path on

this combo lists files but not executables:
clink set exec.enable off
clink set exec.path on

**Comment by chrisant996** [comment]
Oh, I think you mean:

- I want completion for the first word to list files in the current directory.
- I want completion for the first word to also list executables found in the %PATH% variable.

Is that what you're saying?

I can add an `exec.files` setting to also list non-executable files.

Or in the meantime you can use `clink set exec.enable 0` and add this Lua script:

**execcomp.lua**
```lua
local gen = clink.generator(1)

local function get_environment_paths()
    return (os.getenv("path") or ""):explode(";") or {}
end

local function get_exec_suffices()
    return (os.getenv("pathext") or ""):explode(";") or {}
end

local function add_path_files(match_builder, pattern)
    for _, f in ipairs(os.globfiles(pattern, true)) do
        match_builder:addmatch({ match = f.name, type = f.type })
    end
end

function gen:generate(line_state, match_builder)
    -- Only applicable when this is the first word of the line.
    if line_state:getwordcount() > 1 or #line_state:getendword() > 0 then return false end

    -- Add files.
    match_builder:addmatches(clink.filematches("*"))

    -- Add executables from %PATH%.  Executables are defined by %PATHEXT%.
    local paths = get_environment_paths()
    local suffices = get_exec_suffices()
    for _, dir in ipairs(paths) do
        for _, suffix in ipairs(suffices) do
            add_path_files(match_builder, path.join(dir, "*"..suffix))
        end
    end
    return true
end
```

**Comment by chrisant996** [comment]
Or type ` ` (space) to use normal file matching.

- <kbd>x</kbd><kbd>Tab</kbd> uses executable matching for "x".
- <kbd>Space</kbd><kbd>x</kbd><kbd>Tab</kbd> uses normal matching for "x".

That's what the `exec.space_prefix` setting is about.

**Comment by mrx23dot** [comment]
Yeah the goal would be to autocomplete files and executables too. 
(save us typing) without thinking (entering space).

I will go with  exec.enable 0 for now until it's officially released.
but so much better than cmd.exe already, cheers!

**Comment by chrisant996** [comment]
It'll probably be released tonight or this weekend.

It was added in 2ad0734f748055c4dcba7d30039d2cad40c52975.

(FWIW, personally, I turn off `exec.path` because I use <kbd>Tab</kbd> = `menu-complete`, which cycles through matches.  And when it includes every executable on the %PATH%, then matching anything becomes very inconvenient for the first word.)

**Comment by pukkandan** [comment]
> And when it includes every executable on the %PATH%, then matching anything becomes very inconvenient for the first word

This is sometimes true, but I personally prefer that over not matching from path at all. Ideally, there would be an option to de-prioritize matching from path. ie; First the matches cycle through (executable?) files in the cd and only then goes to PATH. Is this a possible feature?

**Comment by chrisant996** [comment]
> This is sometimes true, but I personally prefer that over not matching from path at all. Ideally, there would be an option to de-prioritize matching from path. ie; First the matches cycle through (executable?) files in the cd and only then goes to PATH. Is this a possible feature?

That isn't possible at this time.  Matches are sorted alphabetically, duplicates are removed, and there isn't a way to supply secondary additional text for sorting purposes (or any other purposes).

**Comment by chrisant996** [comment]
Addressed by 2ad0734f748055c4dcba7d30039d2cad40c52975 and 56fa679a6ea46a11c3e9a6c20c4865e1c824b151.

Also -- if there are just certain extensions you want to include, and not all files, then consider updating `%PATHEXT%` to include the additional executable extensions.
